flink/v2.0/flink/src/main/java/org/apache/iceberg/flink/sink/RowDataTransformer.java

```diff
@@ -0,0 +1,545 @@
+        PositionalGetter<?, ?> valueGetter,
+        LogicalType keyType,
+        LogicalType valueType) {
+      MapDataAccessor mapDataAccessor =
+          new MapDataAccessor(
+              new ArrayDataAccessor((PositionalGetter<?, ArrayData>) keyGetter),
+              new ArrayDataAccessor((PositionalGetter<?, ArrayData>) valueGetter));
+      boolean isInListFlag = Boolean.TRUE.equals(isInList.peek());
+      return (data, pos) -> {
+        if (isInListFlag) {
+          mapDataAccessor.key.arrayData = ((ArrayData) data).getMap(pos).keyArray();
+          mapDataAccessor.value.arrayData = ((ArrayData) data).getMap(pos).valueArray();
+        } else {
+          mapDataAccessor.key.arrayData = ((RowData) data).getMap(pos).keyArray();
+          mapDataAccessor.value.arrayData = ((RowData) data).getMap(pos).valueArray();
+        }
+
+        return mapDataAccessor;
+      };
+    }
+
+    @Override
+    public PositionalGetter<?, ?> primitive(Type.PrimitiveType type, LogicalType logicalType) {
+      if (Boolean.TRUE.equals(isInList.peek())) {
+        return arrayPrimitive(type, logicalType);
+      } else {
+        return rowDataPrimitive(type, logicalType);
+      }
+    }
+
+    private PositionalGetter<?, RowData> rowDataPrimitive(Type type, LogicalType logicalType) {
+      switch (type.typeId()) {
+        case BOOLEAN:
+          return RowData::getBoolean;
+        case INTEGER:
+          if (logicalType.getTypeRoot() == LogicalTypeRoot.TINYINT) {
+            return (row, pos) -> (int) row.getByte(pos);
+          } else if (logicalType.getTypeRoot() == LogicalTypeRoot.SMALLINT) {
+            return (row, pos) -> (int) row.getShort(pos);
+          }
+
+          return RowData::getInt;
+        case LONG:
+          return RowData::getLong;
+        case FLOAT:
+          return RowData::getFloat;
+        case DOUBLE:
+          return RowData::getDouble;
+        case STRING:
+          return RowData::getString;
+        case DECIMAL:
+          DecimalType decimalType = (DecimalType) logicalType;
+          return (row, pos) ->
+              row.getDecimal(pos, decimalType.getPrecision(), decimalType.getScale());
+        case DATE:
+        case TIME:
+          return RowData::getInt;
+        case TIMESTAMP:
+        case TIMESTAMP_NANO:
+          switch (logicalType.getTypeRoot()) {
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+              return (row, pos) ->
+                  row.getTimestamp(pos, ((TimestampType) logicalType).getPrecision());
+
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+              return (row, pos) ->
+                  row.getTimestamp(pos, ((LocalZonedTimestampType) logicalType).getPrecision());
+            default:
+              throw new UnsupportedOperationException(
+                  "Not supported logical type: " + logicalType.getTypeRoot());
+          }
+        case UUID:
+        case BINARY:
+        case FIXED:
+          return RowData::getBinary;
```

Comment:
Would it make sense to separate the boilerplate mapping code from the custom type mapping that we see here? It is hard to understand what actually changed. Ideally, we would only supply mappings for the types that changed and leave all others untouched.

Reply:
Let's see what Ryan thinks when he sees these classes.
If we decide to move forward with the transformation layer we can definitely talk about the ways we can make this code a bit more easy to read.